server/src/main/java/org/opensearch/search/aggregations/bucket/terms/MultiTermsAggregator.java

```diff
@@ -213,8 +229,114 @@ public InternalAggregation buildEmptyAggregation() {
+        // For example, if we have a composite key of (a, b) where a is from field1 & b is from field2
+        // We can a find all the composite buckets by visiting both the posting lists
+        // and counting all the documents that intersect for each composite bucket.
+        // This is much faster than visiting the doc values for each document.
+
+        if (weight == null || weight.count(ctx) != ctx.reader().maxDoc()) {
+            // Weight not assigned - cannot use this optimization
+            // weight.count(ctx) == ctx.reader().maxDoc() implies there are no deleted documents and
+            // top-level query matches all docs in the segment
+            return null;
+        }
+
+        String field1, field2;
+        // Restricting the number of fields to 2 and only keyword fields with FieldData available
+        if (this.valuesSources.size() == 2
+            && this.valuesSources.get(0) instanceof ValuesSource.Bytes.WithOrdinals.FieldData
+            && this.valuesSources.get(1) instanceof ValuesSource.Bytes.WithOrdinals.FieldData) {
+            field1 = ((ValuesSource.Bytes.WithOrdinals.FieldData) valuesSources.get(0)).getIndexFieldName();
+            field2 = ((ValuesSource.Bytes.WithOrdinals.FieldData) valuesSources.get(1)).getIndexFieldName();
+
+        } else {
+            return null;
+        }
+
+        Terms segmentTerms1 = ctx.reader().terms(field1);
+        Terms segmentTerms2 = ctx.reader().terms(field2);
+
+        // TODO in this PR itself in coming commits:
+        // 1/ add check for fields cardinality - this might be ineffective for very high cardinality
+        // 2/ check for filter applied or not as default implementation might be resolving it as part of aggregation
+
+        TermsEnum segmentTermsEnum1 = segmentTerms1.iterator();
+
+        while (segmentTermsEnum1.next() != null) {
+            TermsEnum segmentTermsEnum2 = segmentTerms2.iterator();
+
+            while (segmentTermsEnum2.next() != null) {
+
+                PostingsEnum postings1 = segmentTermsEnum1.postings(null);
+                postings1.nextDoc();
+
+                PostingsEnum postings2 = segmentTermsEnum2.postings(null);
+                postings2.nextDoc();
+
+                int bucketCount = 0;
+
+                while (postings1.docID() != PostingsEnum.NO_MORE_DOCS && postings2.docID() != PostingsEnum.NO_MORE_DOCS) {
+
+                    // Count of intersecting docs to get number of docs in each bucket
+                    if (postings1.docID() == postings2.docID()) {
+                        bucketCount++;
+                        postings1.nextDoc();
+                        postings2.nextDoc();
+                    } else if (postings1.docID() < postings2.docID()) {
+                        postings1.advance(postings2.docID());
+                    } else {
+                        postings2.advance(postings1.docID());
+                    }
+                }
```

Comment:
we need to optimize this method.
could you create a fixedbitset and use intersectionCount https://github.com/apache/lucene/blob/ebea2e1492c95b5d6b1e1032485598f901bda286/lucene/core/src/java/org/apache/lucene/util/FixedBitSet.java#L74

Reply:
Agreed. The complexity of intersection logic is highly dependent on the documents in the posting lists. With larger datasets and higher cardinality, the leapfrogging method for intersection evaluation would require more frequent iterations over these lists, which can be expensive.